app/vue/src/client/preview/index.js

```diff
@@ -1,30 +1,72 @@
 import { start } from '@storybook/core/client';
+import Vue from 'vue';
 
 import './globals';
 import render from './render';
 
-const createWrapperComponent = Target => ({
-  functional: true,
-  render(h, c) {
-    return h(Target, c.data, c.children);
-  },
-});
-const decorateStory = (getStory, decorators) =>
-  decorators.reduce(
+export const WRAPS = 'STORYBOOK_WRAPS';
+export const VALUES = 'STORYBOOK_VALUES';
+
+function extractProps(component) {
+  return Object.entries(component.options.props || {})
+    .map(([name, def]) => ({ [name]: def.default }))
+    .reduce((wrap, prop) => ({ ...wrap, ...prop }), {});
+}
+
+function prepare(rawStory, innerStory) {
+  let story = rawStory;
+  // eslint-disable-next-line no-underscore-dangle
+  if (!story._isVue) {
+    if (typeof story === 'string') {
+      story = { template: story };
+    }
+    if (innerStory) {
+      story.components = { ...(story.components || {}), story: innerStory };
+    }
+    story = Vue.extend(story);
+  } else if (story.options[WRAPS]) {
+    return story;
+  }
+
+  return Vue.extend({
+    [WRAPS]: story,
+    [VALUES]: { ...(innerStory ? innerStory.options[VALUES] : {}), ...extractProps(story) },
+    functional: true,
+    render(h, { data, parent, children }) {
+      return h(
+        story,
+        {
+          ...data,
+          props: { ...(data.props || {}), ...(parent.$root[VALUES] || {}) },
+        },
+        children
+      );
+    },
+  });
+}
+
+function decorateStory(getStory, decorators) {
+  return decorators.reduce(
     (decorated, decorator) => context => {
-      const story = () => decorated(context);
-      let decoratedStory = decorator(story, context);
+      let story;
+      const decoratedStory = decorator(() => {
+        story = decorated(context);
+        return story;
+      }, context);
```

Comment:
previously i had the following:
```js
const story = decorated(context);
const decoratedStory = decorator(() => story, context);
```
but this does not work, because the knobs addon (and maybe some others) rly on sideeffects happening during the `getStory` call. this could be improved in the future.